README
NAME
    App::git::ship - Git command for shipping your project

VERSION
    0.02

DESCRIPTION
    App::git::ship is a "git" command for shipping your project to CPAN or
    some other repository.

    This project can also "init" (create) a new project, just "build"
    (prepare for shipping) and "clean" projects.

    The main focus is to automate away the boring steps, but at the same
    time not get in your way. Problems should be solved with sane defaults
    according to standard rules instead of enforcing more rules.

SYNOPSIS
  For end user
      $ git ship -h

      # make a new project

  For developer
      package App::git::ship::some_language;
      use App::git::ship -base;

      # define attributes
      has some_attribute => sub {
        my $self = shift;
        return "default value";
      };

      # override the methods defined in App::git::ship
      sub build {
        my $self = shift;
      }

      1;

TODO
    This project is currently in the EXPERIMENTAL phase, where I'm testing
    things in the real world.

    *   Add a manual for how to make, build and ship a Perl project.

    *   Write down what I think is the key difference between this project
        and the competing projects.

    *   Make a blogpost

    *   Make git aliases for .gitconfig

ATTRIBUTES
  config
      $hash_ref = $self->config;

    Holds the configuration from end user. The config is by default read
    from ".ship.conf" in the root of your project.

  next_version
      $str = $self->next_version;

    Holds the next version to "ship".

  project_name
      $str = $self->project_name;

    Holds the name of the current project. This attribute can be read from
    "config".

  repository
      $str = $self->repository;

    Returns the URL to the first repository that point to "origin". This
    attribute can be read from "config".

  silent
      $bool = $self->silent;
      $self = $self->silent($bool);

    Set this to true if you want less logging. By default it silent is
    false.

METHODS
  abort
      $self->abort($str);
      $self->abort($format, @args);

    Will abort the application run with an error message.

  attr
      $class = $class->attr($name => sub { my $self = shift; return $default_value });

    or ...

      use App::git::ship -base;
      has $name => sub { my $self = shift; return $default_value };

    Used to create an attribute with a lazy builder.

  build
    This method builds the project. The default behavior is to "abort". Need
    to be overridden in the subclass.

  can_handle_project
      $bool = $class->can_handle_project($file);

    This method is called by "detect" in App::git::ship and should return
    boolean true if this module can handle the given git project.

    This is a class method which gets a file as input to detect or have to
    auto-detect from current working directory.

  detect
      $class = $self->detect;
      $class = $self->detect($file);

    Will detect the module which can be used to build the project. This can
    be read from the "class" key in "config" or will in worse case default
    to App::git::ship.

  init
    This method is called when initializing the project. The default
    behavior is to populate "config" with default data:

    *   bugtracker

        URL to the bug tracker. Will be the the "repository" URL without
        ".git", but with "/issues" at the end instead.

    *   homepage

        URL to the project homepage. Will be the the "repository" URL,
        without ".git".

    *   license

        The name of the license. Default to artistic_2
        <http://www.opensource.org/licenses/artistic-license-2.0>.

        See "license" in CPAN::Meta::Spec for alternatives.

  new
      $self = $class->new(%attributes);

    Creates a new instance of $class.

  render
      $self->render($file, \%args);

    Used to render a template by the name $file to a $file. The template
    need to be defined in the "DATA" section of the current class or one of
    the super classes.

  ship
    This method ships the project to some online repository. The default
    behavior is to make a new tag and push it to "origin".

  system
      $self->system($program, @args);

    Same as perl's "system()", but provides error handling and logging.

  import
      use App::git::ship;
      use App::git::ship -base;

    Called when this class is used. It will automatically enable strict,
    warnings, utf8 and Perl 5.10 features.

    "-base" will also make sure the calling class inherit from
    App::git::ship and gets the has function.

SEE ALSO
    *   Dist::Zilla

        This project can probably get you to the moon.

    *   Minilla

        This looks really nice for shipping your project. It has the same
        idea as this distribution: Guess as much as possible.

    *   Shipit

        One magical tool for doing it all in one bang.

COPYRIGHT AND LICENSE
    Copyright (C) 2014, Jan Henning Thorsen

    This program is free software, you can redistribute it and/or modify it
    under the terms of the Artistic License version 2.0.

AUTHOR
    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Changelog for App::git::ship


0.02     Sun Jul  6 14:41:27 2014
       * Need to run ship twice: One for build, and one for ship
       * Use the default MANIFEST.SKIP file in addition to custom rules
       * Add "clean" for Perl project
       * Add Makefile.PL and README to git repository on build
       * Add EXE_FILES to Makefile.PL
       * Fix MANIFEST.SKIP to make a proper release

0.01     Sat Jul  5 15:16:30 2014
       * Add basic init, build and ship for Perl project
       * Add basic init for any project
       * Add detection for Perl projects



lib/App/git/ship/perl.pm
package App::git::ship::perl;

=head1 NAME

App::git::ship::perl - Ship your Perl module

=head1 DESCRIPTION

L<App::git::ship::perl> is a module that can ship your Perl module.

See L<App::git::ship/SYNOPSIS>

=cut

use App::git::ship -base;
use Cwd ();
use File::Basename qw( dirname basename );
use File::Path 'make_path';
use File::Spec;
use Module::CPANfile;

my $VERSION_RE = qr{\b\d+\.[\d_]+\b};

=head1 ATTRIBUTES

=head2 main_module_path

  $str = $self->main_module_path;

Tries to guess the path to the main module in the repository. This is done by
looking at the repo name and try to find a file by that name. Example:

  ./my-cool-project/.git
  ./my-cool-project/lib/My/Cool/Project.pm

This guessing is case-insensitive.

Instead of guessing, you can put "main_module_path" in the config file.

=cut

has main_module_path => sub {
  my $self = shift;
  return $self->config->{main_module_path} if $self->config->{main_module_path};

  my @path = split /-/, basename(Cwd::getcwd);
  my $path = 'lib';
  my @name;

  PATH_PART:
  for my $p (@path) {
    opendir my $DH, $path or $self->abort("Cannot find project name from $path: $!");

    for my $f (readdir $DH) {
      $f =~ s/\.pm$//;
      next unless lc $f eq lc $p;
      push @name, $f;
      $path = File::Spec->catdir($path, $f);
      next PATH_PART;
    }
  }

  return "$path.pm";
};

=head2 project_name

  $str = $self->project_name;

Tries to figure out the project name from L</main_module_path> unless the
L</project_name> is specified in config file.

Example result: "My::Perl::Project".

=cut

has project_name => sub {
  my $self = shift;

  return $self->config->{project_name} if $self->config->{project_name};

  my @name = File::Spec->splitdir($self->main_module_path);
  shift @name if $name[0] eq 'lib';
  $name[-1] =~ s!\.pm$!!;
  join '::', @name;
};

has _cpanfile => sub { Module::CPANfile->load; };

=head1 METHODS

=head2 build

Used to build a Perl distribution.

=cut

sub build {
  my $self = shift;

  $self->clean(0);
  $self->system(prove => split /\s/, $self->config->{build_test_options}) if $self->config->{build_test_options};
  $self->clean(0);
  $self->_render_makefile_pl;
  $self->_timestamp_to_changes;
  $self->_update_version_info;
  $self->system(sprintf '%s %s > %s', 'perldoc -tT', $self->main_module_path, 'README');
  $self->_make('manifest');
  $self->_make('dist');
  $self;
}

=head2 can_handle_project

See L<App::git::ship/can_handle_project>.

=cut

sub can_handle_project {
  my ($class, $file) = @_;
  my $can_handle_project = 0;

  if ($file) {
    return $file =~ /\.pm$/ ? 1 : 0;
  }
  if (-d 'lib') {
    File::Find::find(sub { $can_handle_project = 1 if /\.pm$/; }, 'lib');
  }

  return $can_handle_project;
}

=head2 clean

Used to clean out build files.

=cut

sub clean {
  my $self = shift;
  my $all = shift // 1;
  my @files = qw( Makefile Makefile.old MANIFEST MYMETA.json MYMETA.yml );

  push @files, qw( Changes.bak META.json META.yml ) if $all;
  $self->_dist_files(sub { push @files, $_; });

  for my $file (@files) {
    next unless -e $file;
    unlink $file or warn "!! rm $file: $!" and next;
    say "\$ rm $file" unless $self->silent;
  }

  return $self;
}

=head2 exe_files

  @files = $self->exe_files;

Returns a list of files in the "bin/" directory that has the executable flag
set.

This method is used to build the C<EXE_FILES> list in C<Makefile.PL>.

=cut

sub exe_files {
  my $self = shift;
  my $BIN;

  return unless opendir $BIN, 'bin';
  return map { "bin/$_" } grep { /^\w/ and -x File::Spec->catfile("bin", $_) } readdir $BIN;
}

=head2 init

Used to generate C<Changes> and C<MANIFEST.SKIP>.

=cut

sub init {
  my $self = shift;

  if (my $file = $_[0]) {
    $file = File::Spec->catfile(lib => $file) unless $file =~ m!^.?lib!;
    $self->config({})->main_module_path($file);
    my $work_dir = lc($self->project_name) =~ s!::!-!gr;
    mkdir $work_dir;
    chdir $work_dir or $self->abort("Could not chdir to $work_dir");
    make_path dirname $self->main_module_path;
    open my $MAINMODULE, '>>', $self->main_module_path or $self->abort("Could not create %s", $self->main_module_path);
  }

  symlink $self->main_module_path, 'README.pod' unless -e 'README.pod';

  $self->SUPER::init(@_);
  $self->render('cpanfile');
  $self->render('Changes');
  $self->render('MANIFEST.SKIP');
  $self->render('t/00-basic.t');
  $self->system(qw( git add cpanfile Changes MANIFEST.SKIP t ));
  $self->system(qw( git commit --amend -C HEAD )) if @_;
  $self;
}

=head2 ship

Use L<App::git::ship/ship> and then push the new release to CPAN
using C<cpan-uploader-http>.

=cut

sub ship {
  my $self = shift;
  my $dist_file = $self->_dist_files(sub { 1 });
  my $uploader;

  require CPAN::Uploader;
  $uploader = CPAN::Uploader->new(CPAN::Uploader->read_config_file);

  unless ($dist_file) {
    $self->build;
    $self->abort("Project built. Run (git ship) to post to CPAN and alien repostitory.");
  }

  $self->system(qw( git add Makefile.PL Changes README ));
  $self->system(qw( git commit -a -m ), $self->_changes_to_commit_message);
  $self->SUPER::ship(@_); # after all the changes
  $uploader->upload_file($dist_file);
  $self->clean;
}

sub _author {
  my ($self, $format) = @_;

  open my $GIT, '-|', qw( git log ), "--format=$format" or $self->abort("git log --format=$format: $!");
  my $author = readline $GIT;
  $self->abort("Could not find any author in git log") unless $author;
  chomp $author;
  warn "[ship::author] $format = $author\n" if DEBUG;
  return $author;
}

sub _changes_to_commit_message {
  my $self = shift;
  my ($version, @message);

  close ARGV; # reset <> iterator
  local @ARGV = qw( Changes );
  while (<>) {
    last if @message and /^($VERSION_RE)\s+/;
    push @message, $_ if @message;
    push @message, $_ and $version = $1 if /^($VERSION_RE)\s+/;
  }

  $message[0] =~ s!.*?\n!Released version $version\n\n!s;
  local $" = '';
  return "@message";
}

sub _dist_files {
  my ($self, $cb) = @_;
  my $name = lc($self->project_name) =~ s!::!-!gr;

  opendir(my $DH, Cwd::getcwd);
  while (readdir $DH) {
    next unless /^$name.*\.tar/i;
    return $_ if $self->$cb;
  }

  return undef;
}

sub _make {
  my ($self, @args) = @_;

  $self->_render_makefile_pl unless -e 'Makefile.PL';
  $self->system(perl => 'Makefile.PL') unless -e 'Makefile';
  $self->system(make => @args);
}

sub _render_makefile_pl {
  my $self = shift;
  my $prereqs = $self->_cpanfile->prereqs;
  my $args = { force => 1 };

  $args->{PREREQ_PM} = $prereqs->requirements_for(qw( runtime requires ))->as_string_hash;

  for my $k (qw( build test )) {
    my $r = $prereqs->requirements_for($k, 'requires')->as_string_hash;
    $args->{BUILD_REQUIRES}{$_} = $r->{$_} for keys %$r;
  }

  $self->render('Makefile.PL', $args);
  $self->system(qw( perl -c Makefile.PL )); # test Makefile.PL
}

sub _timestamp_to_changes {
  my $self = shift;
  my $date = localtime;

  local @ARGV = qw( Changes );
  local $^I = '';
  while (<>) {
    $self->next_version($1) if s/^($VERSION_RE)\s*$/{ sprintf "\n%-7s  %s\n", $1, $date }/e;
    print; # print back to same file
  }

  say '# Building version ', $self->next_version unless $self->silent;
  $self->abort('Unable to add timestamp to ./Changes') unless $self->next_version;
}

sub _update_version_info {
  my $self = shift;
  my $version = $self->next_version or $self->abort('Internal error: Are you sure Changes has a timestamp?');
  my %r;

  local @ARGV = ($self->main_module_path);
  local $^I = '';
  while (<>) {
    $r{pod} ||= s/$VERSION_RE/$version/ if /^=head1 VERSION/ .. $r{pod} && /^=(cut|head1)/ || eof;
    $r{var} ||= s/((?:our)?\s*\$VERSION)\s*=.*/$1 = '$version';/;
    print; # print back to same file
  }

  $self->abort('Could not update VERSION in %s', $self->main_module_path) unless $r{var};
}

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014, Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut

1;

__DATA__
@@ .gitignore
~$
*.bak
*.old
*.swp
/blib/
/cover_db
/inc/
/local
/Makefile
/Makefile.old
/MANIFEST$
/MANIFEST.bak
/META*
/MYMETA*
/pm_to_blib
@@ cpanfile
# You can install this projct with curl -L http://cpanmin.us | perl - <%= $_[0]->repository =~ s!\.git$!!r %>/archive/master.tar.gz
requires "perl" => "5.10.0";
test_requires "Test::More" => "0.88";
@@ Changes
Changelog for <%= $self->project_name %>

0.01
       * Started project

@@ Makefile.PL
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
use ExtUtils::MakeMaker;
WriteMakefile(
  NAME => '<%= $_[0]->project_name %>',
  AUTHOR => '<%= $_[0]->_author('%an <%ae>') %>',
  LICENSE => '<%= $_[0]->config->{license} %>',
  ABSTRACT_FROM => '<%= $_[0]->main_module_path %>',
  VERSION_FROM => '<%= $_[0]->main_module_path %>',
  EXE_FILES => [qw( <%= join ' ', $_[0]->exe_files %> )],
  META_MERGE => {
    resources => {
      bugtracker => '<%= $_[0]->config->{bugtracker} %>',
      homepage => '<%= $_[0]->config->{homepage} %>',
      repository => '<%= $_[0]->repository %>',
    },
  },
  BUILD_REQUIRES => <%= $_[1]->{BUILD_REQUIRES} %>,
  PREREQ_PM => <%= $_[1]->{PREREQ_PM} %>,
  test => { TESTS => 't/*.t' },
);
@@ MANIFEST.SKIP
#!include_default
\.swp$
^local/
^MANIFEST\.SKIP
^README\.pod
@@ t/00-basic.t
use Test::More;
use File::Find;

if(($ENV{HARNESS_PERL_SWITCHES} || '') =~ /Devel::Cover/) {
  plan skip_all => 'HARNESS_PERL_SWITCHES =~ /Devel::Cover/';
}
if(!eval 'use Test::Pod; 1') {
  *Test::Pod::pod_file_ok = sub { SKIP: { skip "pod_file_ok(@_) (Test::Pod is required)", 1 } };
}
if(!eval 'use Test::Pod::Coverage; 1') {
  *Test::Pod::Coverage::pod_coverage_ok = sub { SKIP: { skip "pod_coverage_ok(@_) (Test::Pod::Coverage is required)", 1 } };
}

find(
  {
    wanted => sub { /\.pm$/ and push @files, $File::Find::name },
    no_chdir => 1
  },
  -e 'blib' ? 'blib' : 'lib',
);

plan tests => @files * 3;

for my $file (@files) {
  my $module = $file; $module =~ s,\.pm$,,; $module =~ s,.*/?lib/,,; $module =~ s,/,::,g;
  ok eval "use $module; 1", "use $module" or diag $@;
  Test::Pod::pod_file_ok($file);
  Test::Pod::Coverage::pod_coverage_ok($module);
}


lib/App/git/ship.pm
package App::git::ship;

=head1 NAME

App::git::ship - Git command for shipping your project

=head1 VERSION

0.02

=head1 DESCRIPTION

L<App::git::ship> is a C<git> command for shipping your project to CPAN or
some other repository.

This project can also L</init> (create) a new project, just L</build> (prepare
for L<shipping|/ship>) and L</clean> projects.

The main focus is to automate away the boring steps, but at the same time not
get in your way. Problems should be solved with sane defaults according to
standard rules instead of enforcing more rules.

=head1 SYNOPSIS

=head2 For end user

  $ git ship -h

  # make a new project

=head2 For developer

  package App::git::ship::some_language;
  use App::git::ship -base;

  # define attributes
  has some_attribute => sub {
    my $self = shift;
    return "default value";
  };

  # override the methods defined in App::git::ship
  sub build {
    my $self = shift;
  }

  1;

=head1 TODO

This project is currently in the EXPERIMENTAL phase, where I'm testing things
in the real world.

=over

=item * Add a manual for how to make, build and ship a Perl project.

=item * Write down what I think is the key difference between this project and the competing projects.

=item * Make a blogpost

=item * Make git aliases for .gitconfig

=back

=cut

use feature ':5.10';
use strict;
use warnings;
use Carp;
use Data::Dumper ();
use File::Basename 'dirname';
use File::Find ();
use File::Path 'make_path';
use File::Spec ();

use constant DEBUG => $ENV{GIT_SHIP_DEBUG} || 0;

our $VERSION = '0.02';

my %DATA;

=head1 ATTRIBUTES

=head2 config

  $hash_ref = $self->config;

Holds the configuration from end user. The config is by default read from
C<.ship.conf> in the root of your project.

=head2 next_version

  $str = $self->next_version;

Holds the next version to L</ship>.

=head2 project_name

  $str = $self->project_name;

Holds the name of the current project. This attribute can be read from
L</config>.

=head2 repository

  $str = $self->repository;

Returns the URL to the first repository that point to "origin".
This attribute can be read from L</config>.

=head2 silent

  $bool = $self->silent;
  $self = $self->silent($bool);

Set this to true if you want less logging. By default it silent is false.

=cut

__PACKAGE__->attr(config => sub {
  my $self = shift;
  my $file = '.ship.conf';
  my $config;

  open my $CFG, '<', $file or $self->abort("Read $file: $!");

  while (<$CFG>) {
    chomp;
    warn "[ship::config] $_\n" if DEBUG == 2;
    m/\A\s*(?:\#|$)/ and next; # comments
    s/\s+\#.*$//; # remove inline comments
    m/^\s*([^\=\s][^\=]*?)\s*=\s*(.*?)\s*$/ or next;
    $config->{$1} = $2;
    warn "[ship::config] $1 = $2\n" if DEBUG;
  }

  return $config;
});

__PACKAGE__->attr(next_version => sub { 0 });

__PACKAGE__->attr(project_name => sub {
  my $self = shift;
  return $self->config->{project_name} if $self->config->{project_name};
  return 'unknown';
});

__PACKAGE__->attr(repository => sub {
  my $self = shift;
  open my $REPOSITORIES, '-|', q( git remote -v ) or $self->abort("git remote -v: $!");
  my ($repository) = grep { /^origin\b.*\bpush\b/ } <$REPOSITORIES>;
  $repository = sprintf 'https://github.com/%s', $repository ? (split /[:\s+]/, $repository)[2] : lc(join '/', (getpwuid $<)[0], $self->project_name =~ s!::!-!gr);
  warn "[ship::repository] $repository\n" if DEBUG;
  $repository;
});

__PACKAGE__->attr(silent => sub { $ENV{GIT_SHIP_SILENT} // 0 });

=head1 METHODS

=head2 abort

  $self->abort($str);
  $self->abort($format, @args);

Will abort the application run with an error message.

=cut

sub abort {
  my ($self, $format, @args) = @_;
  my $message = @args ? sprintf $format, @args : $format;

  Carp::confess("!! $message") if DEBUG;
  die "!! $message\n";
}

=head2 attr

  $class = $class->attr($name => sub { my $self = shift; return $default_value });

or ...

  use App::git::ship -base;
  has $name => sub { my $self = shift; return $default_value };

Used to create an attribute with a lazy builder.

=cut

sub attr {
  my ($self, $name, $default) = @_;
  my $class = ref $self || $self;
  my $code = "";

  $code .= "package $class; sub $name {";
  $code .= "return \$_[0]->{$name} if \@_ == 1 and exists \$_[0]->{$name};";
  $code .= "return \$_[0]->{$name} = \$_[0]->\$default if \@_ == 1;";
  $code .= "\$_[0]->{$name} = \$_[1] if \@_ == 2;";
  $code .= '$_[0];}';

  eval "$code;1" or die "$code: $@";

  return $self;
}

=head2 build

This method builds the project. The default behavior is to L</abort>.
Need to be overridden in the subclass.

=cut

sub build {
  $_[0]->abort('build() is not available for %s', ref $_[0]);
}

=head2 can_handle_project

  $bool = $class->can_handle_project($file);

This method is called by L<App::git::ship/detect> and should return boolean
true if this module can handle the given git project.

This is a class method which gets a file as input to detect or have to
auto-detect from current working directory.

=cut

sub can_handle_project { $_[1] ? 0 : 1 }

=head2 detect

  $class = $self->detect;
  $class = $self->detect($file);

Will detect the module which can be used to build the project. This
can be read from the "class" key in L</config> or will in worse
case default to L<App::git::ship>.

=cut

sub detect {
  my $self = shift;
  my $file = 'auto-detect';

  if (!@_ and $self->config->{class}) {
    my $class = $self->config->{class};
    eval "require $class;1" or $self->abort("Could not load $class: $@");
    return $class;
  }
  else {
    $file = shift;
    require Module::Find;
    for my $class (sort { length $b <=> length $a } Module::Find::findallmod(__PACKAGE__)) {
      eval "require $class;1" or next;
      next unless $class->can('can_handle_project');
      warn "[ship::detect] $class->can_handle_project($file)\n" if DEBUG;
      return $class if $class->can_handle_project($file);
    }
  }

  $self->abort("Could not figure out what kind of project this is from '$file'");
}

=head2 init

This method is called when initializing the project. The default behavior is
to populate L</config> with default data:

=over 4

=item * bugtracker

URL to the bug tracker. Will be the the L</repository> URL without ".git", but
with "/issues" at the end instead.

=item * homepage

URL to the project homepage. Will be the the L</repository> URL, without ".git".

=item * license

The name of the license. Default to L<artistic_2|http://www.opensource.org/licenses/artistic-license-2.0>.

See L<CPAN::Meta::Spec/license> for alternatives.

=back

=cut

sub init {
  my $self = shift;

  if (@_ and ref($self) eq __PACKAGE__) {
    return $self->detect($_[0])->new($self)->init(@_);
  }

  $self->config({}); # make sure repository() does not die
  $self->system(qw( git init )) unless -d '.git' and @_;
  $self->render('.ship.conf', { homepage => $self->repository =~ s!\.git$!!r });
  $self->render('.gitignore');
  $self->system(qw( git add . ));
  $self->system(qw( git commit -m Initialized )) if @_;
  delete $self->{config}; # regenerate config from .ship.conf
  $self;
}

=head2 new

  $self = $class->new(%attributes);

Creates a new instance of C<$class>.

=cut

sub new {
  my $class = shift;
  my $self = bless @_ ? @_ > 1 ? {@_} : {%{$_[0]}} : {}, ref $class || $class;

  open $self->{STDOUT}, '>&STDOUT';
  open $self->{STDERR}, '>&STDERR';

  $self;
}

=head2 render

  $self->render($file, \%args);

Used to render a template by the name C<$file> to a C<$file>. The template
need to be defined in the C<DATA> section of the current class or one of
the super classes.

=cut

sub render {
  my ($self, $name, $args) = @_;
  my $file = File::Spec->catfile(split '/', $name);
  my $class = ref $self;
  my $str;

  if (-e $file and !$args->{force}) {
    say "# $file exists" unless $self->silent;
    return $self;
  }

  no strict 'refs';
  for my $c ($class, @{"$class\::ISA"}) {
    $str = $c->_data->{$name} and last;
  }

  $self->abort("Could not find template for $name") unless $str;

  local @_ = ($self, $args);
  $str =~ s!<%=(.+?)%>!{
            my $x = eval $1 // die DEBUG ? "($1) => $@" : $@;
            ref $x ? Data::Dumper->new([$x])->Indent(1)->Terse(1)->Sortkeys(1)->Dump : $x;
          }!sge;

  return $str if $args->{to_string};
  make_path dirname($file) or $self->abort("Could not make directory for $file") unless -d dirname $file;
  open my $FH, '>', $file or $self->abort("Could not write $name to $file $!");
  print $FH $str;
  say "# Generated $file" unless $self->silent;
}

=head2 ship

This method ships the project to some online repository. The default behavior
is to make a new tag and push it to "origin".

=cut

sub ship {
  my $self = shift;

  $self->system(qw( git tag ) => $self->next_version);
  $self->system(qw( git push --tags origin ));
}

=head2 system

  $self->system($program, @args);

Same as perl's C<system()>, but provides error handling and logging.

=cut

sub system {
  my ($self, $program, @args) = @_;
  my $exit_code;

  if ($self->silent) {
    open STDERR, '>', File::Spec->devnull;
    open STDOUT, '>', File::Spec->devnull;
  }
  else {
    my $log = "$program @args";
    $log =~ s!\n\r?!\\n!g;
    say "\$ $log" unless $self->silent;
  }

  system $program => @args;
  $exit_code = $? >> 8;

  if ($self->silent) {
    open STDERR, '>&', $self->{STDERR};
    open STDOUT, '>&', $self->{STDOUT};
  }

  $self->abort("'$program @args' failed: $exit_code") if $exit_code;
  $self;
}

=head2 import

  use App::git::ship;
  use App::git::ship -base;

Called when this class is used. It will automatically enable L<strict>,
L<warnings>, L<utf8> and Perl 5.10 features.

C<-base> will also make sure the calling class inherit from
L<App::git::ship> and gets the L<has|/attr> function.

=cut

sub import {
  my ($class, $arg) = @_;
  my $caller = caller;

  if ($arg and $arg eq '-base') {
    no strict 'refs';
    push @{"${caller}::ISA"}, __PACKAGE__;
    *{"${caller}::has"} = sub { attr($caller, @_) };
    *{"${caller}::DEBUG"} = \&DEBUG;
  }

  feature->import(':5.10');
  strict->import;
  warnings->import;
}

# Taken from Mojo::Loader
sub _data {
  my $class = shift;

  return $DATA{$class} if $DATA{$class};
  my $handle = do { no strict 'refs'; \*{"${class}::DATA"} };
  return {} unless fileno $handle;
  seek $handle, 0, 0;
  my $data = join '', <$handle>;

  # Ignore everything before __DATA__ (Windows will seek to start of file)
  $data =~ s/^.*\n__DATA__\r?\n/\n/s;

  # Ignore everything after __END__
  $data =~ s/\n__END__\r?\n.*$/\n/s;

  # Split files
  (undef, my @files) = split /^@@\s*(.+?)\s*\r?\n/m, $data;

  # Find data
  my $all = $DATA{$class} = {};
  while (@files) {
    my $name = shift @files;
    $all->{$name} = shift @files;
  }

  return $all;
}

=head1 SEE ALSO

=over

=item * L<Dist::Zilla>

This project can probably get you to the moon.

=item * L<Minilla>

This looks really nice for shipping your project. It has the same idea as
this distribution: Guess as much as possible.

=item * L<Shipit>

One magical tool for doing it all in one bang.

=back

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014, Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut

1;

__DATA__
@@ .ship.conf
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
class = <%= ref $_[0] %>
project_name = 
homepage = <%= $_[1]->{homepage} %>
bugtracker = <%= join('/', $_[1]->{homepage}, 'issues') =~ s!(\w)//!$1/!r %>
license = artistic_2
build_test_options = # Example: -l -j8
@@ .gitignore
~$
*.bak
*.old
*.swp
/local
@@ test
test = <%= $_[1]->{x} %> + <%= $_[0]->can_handle_project %>.
